### ISSUE TITLE
Floating point precision

### ISSUE THREAD

**Issue by TomTomRixRix** [rgcda/haarpsi#4]
Thanks for providing implementations of HaarPSI. I've stumbled across a tiny difference in the tensorflow implementation compared to the numpy and matlab implementations. The tensorflow implementation uses a 32 bit (single) floating point precision, whereas the other implementations use 64 bit (double). I was wondering if there is any particular reason for this difference? Online I can find reimplementations of HaarPSI enforcing single precision, which leads to (slightly) different scores. Now I'm wondering which precision the original definition used and what therefore would be considered as "correct".

**Comment by lecode-official** [comment]
Yes, the reason why I used 32-bit floating points was that TensorFlow’s optimizers only support 32-bit (at least they did back then, maybe they have added 64-bit support since) and I was using HaarPSI in a loss function. Generally, using 64-bit is preferable due to higher precision and the original implementation (which is the MATLAB implementation) also uses 64-bit and could therefore be considered the “correct” implementation. But, since floating points only approximate real numbers and round to the nearest representable number, the order of operation influences the result, i.e., performing the exact same computation with different orders of operation can yield different results. This is especially true for GPUs, as they perform just-in-time optimizations that may change the order of operation on a machine code level. Also, the underlying implementations of the algorithms are highly optimized and even benchmark different algorithms to select the most efficient one. This means that the same order of operation in Python can result in different results on different NVIDIA GPUs or with different driver/library versions. Considering this, there is almost no way of creating two implementations that yield exactly the same result. That being said, when I implemented the TensorFlow version I checked the results against the MATLAB implementation and ensured that the results had at least a couple of identical significant figures after the decimal point (it has been a couple of years since then, but I believe that I ensured at least 4 significant figures after the decimal point were identical). So for all intents and purposes, this should be close enough to “identical” for any kind of application, unless you have some specific requirements. At the end of the day, this is a philosophical question, as the only “correct” way to compute the HaarPSI index is the mathematical formulation in the paper, which can never be done on computers, as real numbers are unrepresentable with finite memory.

**Comment by TomTomRixRix** [comment]
Thanks a lot for your quick reply. Now I fully understand the decision to implement it in tensorflow with 32 bit. As I'm now interested in using it as a loss function, I'll go for a 64-bit implementation then. I've also noticed in my comparison that the first 4 decimal places are the same. I agree, in most use cases these minor deviations don't make any difference and can therefore be neglected.

**Comment by lecode-official** [comment]
I am happy that I could be of help!
